### PR TITLE
Remove dependency for raft-rs and kvproto

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

I think the repo raft-engine shall be a based library and could be used by other developer even if they does not use `raft-rs` (maybe some other raftlib) and `kvproto`.  I will implement the trait `RaftEngine` in `tikv:component/raft_engine` and define the trait `RaftEngine` in `tikv:component/engine_traits`. Their dependencies will be as following:
```
tikv ----> component/engine_traits, component/raft_engine
component/raft_engine --> raft-engine, component/engine_traits, kvproto, raft-rs
component/engine_rocks --> component/engine_traits, kvproto, raft-rs
```

### CODE REVIEW

File: src/engine.rs [hunk 1]
```diff
@@ -372,12 +374,12 @@ impl FileEngine {
         Ok(bytes)
     }
 
-    fn rewrite(
+    pub fn rewrite(
```
Comment (hicqu) [inline src/engine.rs:377]: It shouldn't be `pub`.

File: src/engine.rs [hunk 2]
```diff
@@ -396,22 +398,22 @@ impl FileEngine {
         Ok(())
     }
 
-    fn sync(&self) -> Result<()> {
+    pub fn sync(&self) -> Result<()> {
         self.pipe_log.sync()
     }
 
-    fn put_msg<M: protobuf::Message>(&self, region_id: u64, key: &[u8], m: &M) -> Result<()> {
+    pub fn put_msg<M: protobuf::Message>(&self, region_id: u64, key: &[u8], m: &M) -> Result<()> {
```
Comment (hicqu) [inline src/engine.rs:405]: Seems there is no `fn put`? It's better to add it.

File: src/engine.rs [hunk 3]
```diff
@@ -733,11 +660,42 @@ impl RaftEngine for FileEngine {
...
+        }
+    }
 
-    fn append_log(engine: &FileEngine, raft: u64, entry: &Entry) {
+    fn append_log(engine: &RaftLogEngine, raft: u64, entry: &RaftEntry) {
         engine.append(raft, vec![entry.clone()]).unwrap();
-        let mut state = RaftLocalState::new();
-        state.last_index = entry.index;
+        let mut state = HardState::new();
+        state.commit = 0;
```
Comment (hicqu) [inline src/engine.rs:698]: I think it's not necessary.

File: src/engine.rs [hunk 4]
```diff
@@ -733,11 +660,42 @@ impl RaftEngine for FileEngine {
...
+    const RAFT_LOG_STATE_KEY: &[u8] = b"R";
+    impl Entry for RaftEntry {
+        fn index(&self) -> u64 {
+            self.get_index()
+        }
+    }
+    pub trait RaftEngine {
+        fn put_raft_state(&self, raft_group_id: u64, state: &HardState) -> Result<()>;
+        fn get_raft_state(&self, raft_group_id: u64) -> Result<Option<HardState>>;
+        fn commit_to(&self, raft_group_id: u64, index: u64) {
```
Comment (hicqu) [inline src/engine.rs:674]: Seems it's not necessary. Could you take a try to remove it?

File: src/engine.rs [hunk 5]
```diff
@@ -707,23 +638,19 @@ impl RaftEngine for FileEngine {
         Ok(will_force_compact)
     }
 
-    fn has_builtin_entry_cache(&self) -> bool {
+    pub fn has_builtin_entry_cache(&self) -> bool {
```
Comment (hicqu) [inline src/engine.rs:641]: It can be removed.

File: src/engine.rs [hunk 6]
```diff
@@ -707,23 +638,19 @@ impl RaftEngine for FileEngine {
...
         true
     }
 
-    fn gc_entry_cache(&self, raft_group_id: u64, to: u64) {
-        self.compact_cache_to(raft_group_id, to)
-    }
-
-    fn flush_stats(&self) -> CacheStats {
+    pub fn flush_stats(&self) -> CacheStats {
```
Comment (hicqu) [inline src/engine.rs:645]: Would you mind to document the method?

**Comment by hicqu** [comment]
rest LGTM. Thanks very much!

File: src/engine.rs (hunk 6 above)
Comment (Little-Wallace) [reply src/engine.rs:645]: Ok

**Review by hicqu — APPROVED** [review]

**Pull request merged**